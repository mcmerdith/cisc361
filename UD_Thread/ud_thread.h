/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);
