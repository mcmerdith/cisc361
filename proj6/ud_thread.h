/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);

/**
 * semaphore
 */

// Create a new semaphore pointed to by sp with a count value of sem_count.
int sem_init(sem_t **sp, int sem_count);

// Current thread does a wait (P) on the specified semaphore.
void sem_wait(sem_t *sp);

// Current thread does a signal (V) on the specified semaphore.
// Follow the Mesa semantics (p. 9 of OSTEP Chapter 30 Condition Variables) where the thread that signals continues,
// and the first waiting (blocked) thread (if there is any) becomes ready.
void sem_signal(sem_t *sp);

// Free any memory related to the specified semaphore.
// If there are queued TCBs, they are moved to the ready queue, and this may imply that the semantics of the application might be incorrect.
void sem_destroy(sem_t **sp);

typedef void sem_t;