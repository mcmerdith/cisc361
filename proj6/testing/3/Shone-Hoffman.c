/*
 * Evan Shone and Jake Hoffman
 * This tests semphores with three different
 * use cases: Synchronization, Locks, Countdown
 */

#include "ud_thread.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

sem_t *sync;
sem_t *lock;
sem_t *count;

int NUM_THREADS = 3; /* less than 10 */
int NUM_ITERATIONS = 3;
int COUNTDOWN_SIZE = 3;

void sync_wait(int id);
void sync_signal(int id);
void lock_thread(int id);
void countdown_thread(int id);
void printBreak();

int main(void)
{
    /* init thread library */
    printBreak();
    printf("About to initialize library\n");
    t_init();
    printf("Library initialized\n");

    /* make semaphores */
    printBreak();
    printf("About to initialize semaphores\n");
    sem_init(&sync, 0);
    sem_init(&lock, 1);
    sem_init(&count, COUNTDOWN_SIZE);
    printf("Semaphores initialized\n");

    /* Sync Test */
    printBreak();
    printf("\tSynchronization Test\n");
    printBreak();

    /* make wait and signal threads */
    printf("Making sync_wait\tthread id:%d\n", 1);
    t_create(sync_wait, 1, 1);
    printf("Making sync_signal\tthread id:%d\n", 2);
    t_create(sync_signal, 2, 1);

    /* yield until sem is signaled */
    int yield_i = 0;
    for (; yield_i < NUM_ITERATIONS; yield_i++)
    {
        printBreak();
        printf("Thread id:main\titer:%d\tyield\n", yield_i + 1);
        t_yield();
    }

    /* yield once more to signal the sem */
    printBreak();
    printf("Thread id:main\t\tyield\n");
    t_yield();
    printf("\n\tsync_wait should now be woken up\n\n");
    printf("Thread id:main\t\tyield\n");
    t_yield();

    printBreak();
    printf("\tSynchronization Test Complete\n");
    printBreak();

    /* Lock Test */
    printBreak();
    printf("\tLock Test\n");
    printBreak();

    int thread_i = 1, lock_i = 0;
    /* Make some threads */
    printf("About to make %d lock threads...\n", NUM_THREADS);
    for (; thread_i < NUM_THREADS + 1; thread_i++)
    {
        printf("main creating thread id:%d\n", thread_i + 2);
        t_create(lock_thread, thread_i + 2, 1);
    }

    /* yield so all the threads run through */
    for (; lock_i < NUM_ITERATIONS * (NUM_THREADS + 2); lock_i++)
    {
        printBreak();
        printf("Thread id:main\titer:%d\tyield\n", lock_i + 1);
        t_yield();
    }

    printBreak();
    printf("\tLock Test Complete\n");
    printBreak();

    /* Countdown Test */
    printBreak();
    printf("\tCountdown Test\n");
    printBreak();

    /* Make some threads */
    printf("About to make %d count threads...\n", NUM_THREADS * 2);
    for (; thread_i < (NUM_THREADS * 3) + 1; thread_i++)
    {
        printf("main creating thread id:%d\n", thread_i + 2);
        t_create(countdown_thread, thread_i + 2, 1);
    }

    printBreak();
    printf("Only %d threads will be able to make it\n", COUNTDOWN_SIZE);
    t_yield();

    printBreak();
    printf("Those threads signal & terminate\n");
    t_yield();

    printBreak();
    printf("The rest of the threads make it\n");
    t_yield();

    printBreak();
    printf("The rest of the threads signal & terminate\n");
    t_yield();

    printBreak();
    printf("\tCountdown Test Complete\n");

    /* destroy the semaphores */
    printBreak();
    printf("About to destroy semaphores\n");
    sem_destroy(&sync);
    sem_destroy(&lock);
    sem_destroy(&count);
    printf("Semaphores destroyed\n");

    t_shutdown();

    printBreak();
    printf("\tTEST COMPLETE\n");
    printBreak();
    return 0;
}

void countdown_thread(int id)
{
    sem_wait(count);
    printf("Thread id:%d\tmade it!\n", id);
    t_yield();
    printf("Thread id:%d\tsignal & terminate\n", id);
    sem_signal(count);
    t_terminate();
}

void lock_thread(int id)
{
    int yield_i = 0;
    printf("Thread id:%d\t\tlock\n", id);
    sem_wait(lock);
    for (; yield_i < NUM_ITERATIONS; yield_i++)
    {
        printf("Thread id:%d\titer:%d\tyield\n", id, yield_i + 1);
        t_yield();
    }
    printf("Thread id:%d\t\tunlock\n", id);
    sem_signal(lock);
    printf("Thread id:%d\titer:%d\tyield\n", id, yield_i + 1);
    t_yield();
    printf("Thread id:%d\t\tterminating...\n", id);
    t_terminate();
}

void sync_wait(int id)
{
    printf("Thread id:%d\t(wait)\twait\n", id);
    sem_wait(sync);
    printf("Thread id:%d\t(wait)\tafter wait\n", id);
    printf("Thread id:%d\t(wait)\tterminating...\n", id);
    t_terminate();
}

void sync_signal(int id)
{
    int yield_i = 0;

    for (; yield_i < NUM_ITERATIONS; yield_i++)
    {
        printf("Thread id:%d\titer:%d\tyield\n", id, yield_i + 1);
        t_yield();
    }

    printf("Thread id:%d\t(sig)\tsignal\n", id);
    sem_signal(sync);
    printf("Thread id:%d\t(sig)\tafter signal\n", id);
    printf("Thread id:%d\t(sig)\tyield\n", id);
    t_yield();
    printf("Thread id:%d\t(sig)\tafter yield\n", id);
    printf("Thread id:%d\t(sig)\tterminating...\n", id);
    t_terminate();
}

void printBreak()
{
    printf("---------------------------------------------------\n");
}