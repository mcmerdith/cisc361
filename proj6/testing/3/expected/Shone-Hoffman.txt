---------------------------------------------------
About to initialize library
Library initialized
---------------------------------------------------
About to initialize semaphores
Semaphores initialized
---------------------------------------------------
	Synchronization Test
---------------------------------------------------
Making sync_wait	thread id:1
Making sync_signal	thread id:2
---------------------------------------------------
Thread id:main	iter:1	yield
Thread id:1	(wait)	wait
Thread id:2	iter:1	yield
---------------------------------------------------
Thread id:main	iter:2	yield
Thread id:2	iter:2	yield
---------------------------------------------------
Thread id:main	iter:3	yield
Thread id:2	iter:3	yield
---------------------------------------------------
Thread id:main		yield
Thread id:2	(sig)	signal
Thread id:2	(sig)	after signal
Thread id:2	(sig)	yield

	sync_wait should now be woken up

Thread id:main		yield
Thread id:1	(wait)	after wait
Thread id:1	(wait)	terminating...
Thread id:2	(sig)	after yield
Thread id:2	(sig)	terminating...
---------------------------------------------------
	Synchronization Test Complete
---------------------------------------------------
---------------------------------------------------
	Lock Test
---------------------------------------------------
About to make 3 lock threads...
main creating thread id:3
main creating thread id:4
main creating thread id:5
---------------------------------------------------
Thread id:main	iter:1	yield
Thread id:3		lock
Thread id:3	iter:1	yield
Thread id:4		lock
Thread id:5		lock
---------------------------------------------------
Thread id:main	iter:2	yield
Thread id:3	iter:2	yield
---------------------------------------------------
Thread id:main	iter:3	yield
Thread id:3	iter:3	yield
---------------------------------------------------
Thread id:main	iter:4	yield
Thread id:3		unlock
Thread id:3	iter:4	yield
---------------------------------------------------
Thread id:main	iter:5	yield
Thread id:4	iter:1	yield
Thread id:3		terminating...
---------------------------------------------------
Thread id:main	iter:6	yield
Thread id:4	iter:2	yield
---------------------------------------------------
Thread id:main	iter:7	yield
Thread id:4	iter:3	yield
---------------------------------------------------
Thread id:main	iter:8	yield
Thread id:4		unlock
Thread id:4	iter:4	yield
---------------------------------------------------
Thread id:main	iter:9	yield
Thread id:5	iter:1	yield
Thread id:4		terminating...
---------------------------------------------------
Thread id:main	iter:10	yield
Thread id:5	iter:2	yield
---------------------------------------------------
Thread id:main	iter:11	yield
Thread id:5	iter:3	yield
---------------------------------------------------
Thread id:main	iter:12	yield
Thread id:5		unlock
Thread id:5	iter:4	yield
---------------------------------------------------
Thread id:main	iter:13	yield
Thread id:5		terminating...
---------------------------------------------------
Thread id:main	iter:14	yield
---------------------------------------------------
Thread id:main	iter:15	yield
---------------------------------------------------
	Lock Test Complete
---------------------------------------------------
---------------------------------------------------
	Countdown Test
---------------------------------------------------
About to make 6 count threads...
main creating thread id:6
main creating thread id:7
main creating thread id:8
main creating thread id:9
main creating thread id:10
main creating thread id:11
---------------------------------------------------
Only 3 threads will be able to make it
Thread id:6	made it!
Thread id:7	made it!
Thread id:8	made it!
---------------------------------------------------
Those threads signal & terminate
Thread id:6	signal & terminate
Thread id:7	signal & terminate
Thread id:8	signal & terminate
---------------------------------------------------
The rest of the threads make it
Thread id:9	made it!
Thread id:10	made it!
Thread id:11	made it!
---------------------------------------------------
The rest of the threads signal & terminate
Thread id:9	signal & terminate
Thread id:10	signal & terminate
Thread id:11	signal & terminate
---------------------------------------------------
	Countdown Test Complete
---------------------------------------------------
About to destroy semaphores
Semaphores destroyed
---------------------------------------------------
	TEST COMPLETE
---------------------------------------------------